feat: add npm audit licenses

bnb opened the pull request:
Begins to implement `npm audit licenses`, which is from https://github.com/npm/rfcs/pull/18.

bnb committed:
feat: add license checking with json output

bnb committed:
test: test for license checking with json output

bnb committed:
deps: move licensee to dep rather than devDep

ljharb commented on lib/audit.js:
```diff
@@ -74,6 +80,51 @@ class Audit extends ArboristWorkspaceCmd {
+      'contributors'
+    ]
+
+    const out = {}
+    for(const dep of dependencies) {
+      out[dep.name] = {}
+      for (const prop of loggableProperties) {
+        out[dep.name][prop] = dep[prop]
+      }
+    }
```
nit:
```suggestion
    const out = Object.fromEntries(dependencies.map((dep) => [
      dep.name,
      Object.fromEntries(loggableProperties.map((prop) => [prop, dep[prop]]),
    ]))
```
altho sadly Object.fromEntries isn't in node 10, so we'd have to add `object.fromentries` to do this :-(

i still enjoyed typing it out to see the clarity improvement

bnb replied:
potentially an enhancement for npm@8 if node@10 support is dropped in it :)

ljharb commented on lib/audit.js:
```diff
@@ -48,10 +49,15 @@ class Audit extends ArboristWorkspaceCmd {
 
   exec (args, cb) {
-    this.audit(args).then(() => cb()).catch(cb)
+    const auditType = this.npm.config.get('audit-type')
+    if (auditType === 'license') {
+      this.auditLicenses(args).then(() => cb()).catch(cb)
+    } else { 
+      this.auditAdvisories(args).then(() => cb()).catch(cb)
+    }
```
does this mean that `npm audit` can only do _either_ licenses or advisories, but not both?

Should `audit-type` be validated somehow so it's only "license" or "advisory"?

bnb replied:
@ljharb in pairing with @ruyadorno today, we intentionally kept this limited to "getting it working" and not fully implementing. This was the minimum way to get it working. There will be individual ways to do advisories, licenses, and both (presuming both will be the default, since `license` is going to be opt-in rather than opt-out).

Ruy's recommendation was `npm audit --audit-type=license` as a more npm-y API, so we went with that. I'm not opposed to other APIs though and presume we'll end up bikeshedding a bit.

ljharb replied:
When I just want one type, that totally makes sense to me - but when I want both, I'd want some config way to per-command override the global `audit-type=license` i plan to configure.

As for the default, I'd assume licenses aren't validated unless there's a config for it, which means it'd be off initially?

bnb replied:
oop, sorry thought I replied to this last night:

the intent is to have `npm audit` do **both** licenses and advisories by default. It can **optionally** do _just_ licenses _or_ advisories.

the implementation here is an intentionally basic one that's the result of @ruyadorno and myself pairing trying to get a _very basic_ implementation working that can be expanded on to include everything that's been scoped out, hence WIP status of the PR.

ljharb replied:
OK - so what `audit-type` value would I provide to override (a global config setting that restricts it to "just licenses")?

bnb replied:
`advisories`: https://github.com/npm/cli/pull/3452/files#diff-2a8ed1f0d31e4c759861aa941a2dc7e4965c10d96a79bad6499301a9a0dadbc2R56

bnb committed:
chore: minor style changes

bnb committed:
feat: some more work on Audit

bnb committed:
chore: add licenses to allow in license check

bnb commented:
Updated this a bit. You can run it with `node . audit --audit-type=license --json" to get JSON output.

I've got user configuration working (based on `audit` in `package.json` - I'm not attached to this, it's just the path I took and I'm happy to revisit later). The API is _slightly_ different than licensee:

- It does not take BlueOak categorization. If we want that, I think it should come in a separate RFC. It would be trivial to add.
- Defaults to correcting assertions (licensee does not), allowing opt-out if end-users want that.
- All properties are on `audit` > `licenses` in `package.json`. This is slightly different than the config of licensee. All the available properties (outside of the aforementioned `blueoak` property) are available, just under the license property.
- What licensee calls `spdx` I have called `identifiers` since that's a more... clear and concise name to those who don't know what SPDX License IDs are. The code in the PR maps `identifiers` to `spdx` when we run `licensee`.

Some additional notes: 
- Added `--audit-type=advisories` that simply duplicates how audit is presently run, with the intent of eventually adding the license checking to the default case. Ideally if the output is JSON, a single object should be returned rather than two different objects. I think this needs to be done outside of the npm/cli code, though... this specific concept leads me to believe there's a larger refactoring that needs to happen here, since it very quickly leads to other questions like "how is this data integrated naturally into non-JSON audit UI" 🙃 
- I've generally done my best to leave comments where I realized additional work was needed. I hope none of them are half-thought out but some might be :P

timhaines commented:
FWIW, the npm docs refer to SPDX - it's something they seek to educate on.  https://docs.npmjs.com/cli/v7/configuring-npm/package-json#license

Perhaps it's worth considering spdx_identifers or similar, to indicate it shouldn't be free-form.

bnb commented:
> Perhaps it's worth considering spdx_identifers or similar, to indicate it shouldn't be free-form.

no other property used by npm uses an underscore - I don't think this is a particularly npm-y approach.

IMO that exact doc you linked is the justification for using `identifiers`. The `license` property _isn't_ `spdx_license_id` or some variant - it's just `license`, despite the fact that it *should* be an SPDX license ID (it realistically doesn't have to be).

ljharb commented:
Does npm even allow publishing anymore with a non-spdx license field?

bnb commented:
You can publish freeform text in that field, yes. If not, it'd prevent things like BlueOak's license from being used which is IMO a very bad thing for open source and progress.

bnb committed:
Merge branch 'latest' into npm-audit-licenses

ljharb commented:
Free form text, but not non-strings?

timhaines commented:
FWIW, Blue Oak's license has an SPDX code; `BlueOak-1.0.0`. But yeah, you do see all kinds of content in that field.  "See license in license" is somewhat common.

bnb commented:
> Free form text, but not non-strings?

AFAIK you can still do an object (the npmjs.com site renders it correctly despite it being deprecated) but you really shouldn't.

bnb commented:
Small update on this: After pairing with @izs it seems like the best approach is to update [Arborist](https://github.com/npm/arborist) (specifically, updating [AuditResolver](https://github.com/npm/arborist/blob/main/lib/audit-report.js#L20) to support licenses in addition to vulnerabilities), since without doing so it's relatively challenging to get nicer bits of `npm audit` that the ecosystem would expect.

I'll probably be a bit slow getting around to that. Happy to pair on it if anyone's interested.

wraithgar commented:
Closing this while folks work on it.  Please do keep your branch in the forked repo. Trying to clear out our pull requests so that they are more manageable.
